compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -4012,15 +4013,56 @@ public void visit(BLangBinaryExpr binaryExpr) {
         checkDecimalCompatibilityForBinaryArithmeticOverLiteralValues(binaryExpr);
 
         SymbolEnv rhsExprEnv;
-        BType lhsType = checkExpr(binaryExpr.lhsExpr, env);
+        BType lhsType;
+        if (binaryExpr.lhsExpr instanceof BLangNumericLiteral) {
```

Comment:
We can use the node kind instead.

```suggestion
        if (binaryExpr.lhsExpr.getKind() == NodeKind.NUMERIC_LITERAL) {
```

We usually try to avoid expensive instanceof checks.